### PR TITLE
hide internal symbols in the goto trace

### PR DESCRIPTION
**Pull request by tautschnig**
None of this is my work - these are commits already merged into develop (test-gen-support).

### CODE REVIEW

File: src/goto-symex/build_goto_trace.cpp [hunk 1]
```diff
@@ -109,6 +109,67 @@ exprt adjust_lhs_object(
...
+{
+  // set internal for dynamic_object in both lhs and rhs expressions
+  set_internal_dynamic_object(SSA_step.ssa_lhs, goto_trace_step, ns);
+  set_internal_dynamic_object(SSA_step.ssa_rhs, goto_trace_step, ns);
+
+  // set internal field to CPROVER functions (e.g., __CPROVER_initialize)
+  if(SSA_step.is_function_call())
+  {
+    if(SSA_step.source.pc->source_location.as_string().empty())
+      goto_trace_step.internal=true;
```
Comment (pkesseli) [inline src/goto-symex/build_goto_trace.cpp]: Perhaps `return` after setting internal to true?

File: src/goto-symex/build_goto_trace.cpp [hunk 2]
```diff
@@ -109,6 +109,67 @@ exprt adjust_lhs_object(
...
+  }
+
+  // set internal field to input and output steps
+  if(goto_trace_step.type==goto_trace_stept::typet::OUTPUT ||
+      goto_trace_step.type==goto_trace_stept::typet::INPUT)
+  {
+    goto_trace_step.internal=true;
```
Comment (pkesseli) [inline src/goto-symex/build_goto_trace.cpp]: Perhaps `return` after setting internal to true?

**Comment by tautschnig** [comment]
To be honest, I'm failing to understand what these commits are trying to achieve. There may be a clue in https://github.com/diffblue/test-gen/issues/20, but it's not public so I can't tell.

If it helps, #709 is the main original PR, but it doesn't have much of a discussion. My main question is: how is "internal" different from "hidden"? @lucasccordeiro @peterschrammel @pkesseli any insights? Thanks!

**Comment by kroening** [comment]
The management of 'hiding steps' on the trace appears to be way to late to me. This PR seems to be a hack to me.

**Comment by lucasccordeiro** [comment]
The UI team requested us to hide the dynamic_object and CPROVER internal functions (e.g., __CPROVER_initialize) from the CBMC counterexample. In order to keep compatibility, we decided to add the ``internal`` field to the JSON output so that the dynamic_object and CPROVER internal functions could be easily hidden in the UI application.

**Comment by tautschnig** [comment]
Thanks a lot for the input. I do not have sufficient state on customers to tell whether anyone does depend on those bits not being hidden and thus call out to @kroening for clarification. One of two things need to happen: either this feature needs to be documented properly to make clear what is "hidden" vs what is "internal", or "internal" should be removed and those bits simply be "hidden." Assigning to @kroening to request guidance.

**Comment by peterschrammel** [comment]
See also https://github.com/diffblue/cbmc/issues/1311

**Comment by peterschrammel** [comment]
I investigated this further:

Current sources of hidden:
* `goto_functionst::type` has a hidden attribute that is set if the function contains a `__CPROVER_hide label`
* symbols with ` #return_value!`  (sic) suffixes are hidden in `symex_assign` 
* assignment steps with source locations with `hide` attribute set are hidden in `symex_assign` (but this seems to be used only in `goto_inline` where the code setting it is deactivated)
* declaration steps are set hidden if the identifier symbol `is_auxiliary` 

Current uses of hidden:
* `hidden` steps are hidden in `show_goto_trace` in the plain text output
* `hidden` steps are hidden in graphml output. 
* `hidden` are not instrumented in `cover` 

The purpose of `hidden` is to hide steps in the trace, 
whereas the purpose of `internal`is to label all lhs symbols that are not known to occur in the user's program. The two notions do not coincide for dynamic objects, for example.

**Comment by kroening** [comment]
Ok, so "internal" is a property of a variable symbol, and not a property of a program step.

Would it not be better to have a namespace for the internal symbols?

Or we may wish to output a symbol table with the trace?

**Comment by tautschnig** [comment]
@peterschrammel Would you be able to take a lead on this issue? This seems to be one that needs careful consideration within Diffblue. (And it is one of the blockers for progress on the develop->master transition.)

**Comment by peterschrammel** [comment]
@tautschnig, yes.

**Comment by peterschrammel** [comment]
@kroening, yes, it's true that `internal` is a property of a symbol rather than a step.
Yet, the symbols we have to know about are exactly those occurring in `lhs` of assignments in the trace (not everything in the symbol table). Therefore it seemed more appropriate to attach the information to the trace steps.

We could output a list of non-internal symbols if this is less contentious.

**Comment by tautschnig** [comment]
Had some further out-of-band discussion with @peterschrammel on this. Taking this onto my stack of tasks in order to clean it up/take it further.

**Comment by tautschnig** [comment]
While the develop->master transition might not happen as originally planned, this PR still includes a significant discussion about "internal" symbols. To be resolved by @peterschrammel and myself.

**Comment by NlightNFotis** [comment]
Hi @tautschnig, is that still something we want to push through?

This looks in a relatively good state (no merge conflicts and approved by other contributors) but it will need some work still (it adds a file to `regression/cbmc-java` that no longer appears to be there (I'm assuming it's under `jbmc/` somewhere).

**Comment by NlightNFotis** [comment]
Also, I just noticed - this is merging to `master`, not `develop`.

Is that still something we want?

**Comment by martin-cs** [comment]
@NlightNFotis the PR or the master branch in general.  I think both are valid questions but I think the later one probably needs an issue of it's own.

**Comment by NlightNFotis** [comment]
@martin-cs I'm talking about both.

1) Do we want to go through with the PR in general? Is this adding value? It looks in a better state than most other old PRs, but I realised that's probably because it's targeting `master` which hasn't seen action in months (probably years) so that's why there's probably no conflicts. 
2) Should we even be targeting `master` with PRs? My understanding is that we are using `develop` as our `master` (sort of) and any merge to `master` should be using `develop` as the source. Is that a fair statement?

By the way I looked into the code changes a bit, and they appear to be replicated in `develop` in any case, so maybe this PR is wholly obsolete?

**Comment by martin-cs** [comment]
1. I think having some way to hide purely internal things in the traces is useful, esp. as they are getting larger.  As the discussion says, it would be best to have 1 way of doing this, not lots.

2. I wasn't at the meeting where the `master` / `develop` split was proposed but my understanding is that the original idea was that develop would move faster and things would be cherry-picked across to master when they were of sufficient quality.  Releases would then be from master.  I think.  `master` has stagnated so much and releases are now from `develop` so ... I don't really see there being much point in having the `master` branch at all.

The comment at the top does suggest it is already in develop.  I guess this might have been part of the "feeding changes back from develop to master" process.  Maybe @tautschnig knows?

**Comment by peterschrammel** [comment]
There was once a plan to merge releases from develop into master so that the head of master is always the last (stable) release version, but we never really started doing that and created releases from develop instead. Atm the master branch is completely stale, so merging PRs into it is pointless.

**Comment by peterschrammel** [comment]
Based on https://github.com/diffblue/cbmc/pull/1272#issuecomment-326919682, https://github.com/diffblue/cbmc/pull/1272#issuecomment-327004442 and https://github.com/diffblue/cbmc/pull/1272#issuecomment-338585420 the management of hiding steps needs a complete overhaul.

**Comment by TGWDB** [comment]
> Based on [#1272 (comment)](https://github.com/diffblue/cbmc/pull/1272#issuecomment-326919682), [#1272 (comment)](https://github.com/diffblue/cbmc/pull/1272#issuecomment-327004442) and [#1272 (comment)](https://github.com/diffblue/cbmc/pull/1272#issuecomment-338585420) the management of hiding steps needs a complete overhaul.

Based on this and the general concerns with this PR I'm closing it. While this does not resolve the question of hiding some `internal` symbols in the trace, since this PR is out of date, badly targeted, and not doing the hiding right anyway, it seems pointless to keep it open. Please create a new PR with a better approach (or create an RFC if you wish to discuss) for addressing the underlying problem in future. (Note that this PR can still be observed/referenced, closing only since it is not going to be merged/continued in it's current state.)

**Pull request closed without merging**